[2단계 - 로또 리팩토링] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
피드백 해주신 내용들 모두 적용하여 수정하였고, 우테코에서 제공해준, 로또 피드백과 제 나름대로의 생각을 담아 코드를 리팩토링했습니다.
최대한 커밋 내용에 코드 변경의 의도를 전달하기위해 노력했습니다.

리뷰 감사합니다 :) 😀

Prolog : https://prolog.techcourse.co.kr/studylogs/3887

changuii committed:
refactor : constans의 enum의 작성 컨벤션을 통일하라

changuii committed:
refactor : NumberGenerator의 이름을 RandomNumberGenerator로 변경하라

changuii committed:
refactor : Lotto 중복 검사 내부 변수의 이름이 set으로 되어있던 것을 해결하라

- set을 removedDuplicate로 변경하였다.

changuii committed:
refactor : LottoMachine의 issueLottos 로직을 stream으로 처리하라

- 불필요한 변수 선언을 제거하여 가독성을 높였다.
- 로또의 생성을 메서드로 분리하여 가독성을 높였다.
- 불변 리스트로 반환하여 변경이 불가능하도록 하였다.

changuii committed:
refactor : WinningNumbers의 중복 검사 변수의 이름을 명확하도록 변경하라

changuii committed:
refactor : LottoNumber의 equals() 메서드 매개변수를 불변으로 만들어라

changuii committed:
refactor : compareTo의 변하지 않는 변수에도 final 키워드를 붙여라

changuii committed:
refactor : Lotto가 당첨 번호와의 매칭 수를 계산할 때, 상태를 모두 넘기지 않고 번호만 넘기도록 하라

- calculateWinningNumbersMatchCount의 메서드를 stream으로 처리하였다.
- containsLottoNumber을 만들어, LottoNumber가 당첨 번호에 존재하는지 여부를 boolean으로 반환하였다.

changuii committed:
test : WinningLotto의 containsLottoNumber를 테스트하라

changuii committed:
refactor : Lotto의 번호 리스트를 전달하는 것이 아닌, WinningNumbers가 bonusBall을 Lotto에게 전달해 비교하도록 하라

changuii committed:
refactor : WinningResult의 계산 로직을 stream으로 변경하라

- 불필요한 변수 선언을 제거하여 가독성을 높였다.

changuii committed:
refactor : WinningResult의 validateRankSize가 indent 요구사항을 만족하게 하라

changuii committed:
refactor : LottoFactory의 객체화 불가능이라는 의도가 쉽게 전달되도록 하라

- abstract 키워드로 객체를 생성하지 못하게 하는 대신, private 생성자로 못하도록 하여 의도를 명확하게 전달하도록 했다.

changuii committed:
refactor : method의 명명을 좀 더 포괄적으로 작성하라

- Lotto의 입장에서는, BonusNumber만, 포함되는 메서드일 필요 없이 LottoNumber라면 무엇이든, 포함되는지 확인할 수 있다.
- 따라서 메서드 명과 매개변수 명을 변경했다.

changuii committed:
refactor : WinningNumbersTest를 리팩토링하라

- TestFixture를 만들어, 중복되는 코드를 제거하였다.
- getter() 대신, extracting()을 활용해 테스트 코드를 위한 getter()를 제거하였다.

changuii committed:
refactor : LottoTest를 리팩토링하라

- TestFixture를 만들어, 중복되는 코드를 줄였다.

changuii committed:
refactor : PurchaseAmount의 테스트를 위한 getter()를 제거하여라

changuii committed:
refactor : LottoNumber의 Comparable의 제네릭을 선언하여라

- compareTo에서, 캐스팅하는 대신, Comparable의 제네릭을 선언하여 바로 LottoNumber로 접근할 수 있도록 하였다.

changuii committed:
refactor : executeWithRetry의 while(true)를 더 직관적으로 수정하라

- 재시도 로직을 무한 루프 대신, 재귀를 통해 관리하여 실수를 줄이도록 하였다.
- depth가 너무 깊어지지 않도록 최대 재시도 횟수를 추가하여 관리하도록 변경하였다.

changuii committed:
docs : 재시도 횟수 10회 발생시, 프로그램 종료 로직에 대한 설명을 추가하라

changuii committed:
refactor : OutputMessage로 출력 메시지를 분리하지 말고, View에서 관리하라

- 매우 밀접하게 연관된 메시지와 출력 로직을 한 곳에 두어 수정이 필요할 때, 이동할 필요가 없도록 했다.
- 중복으로 나타나는 if문은 selectByFlag 메서드로 묶어 내었다.

changuii committed:
refactor : LottoRank와 연관된 출력을 메서드로 분리하여라

- printLottoRanks로 메서드를 추출하여 printLottosStatistics 메서드를 읽을 때의 가독성을 높였다.

changuii committed:
refactor : flag보다 의미있는 변수명으로 명명하라

changuii committed:
refactor : isDamage()처럼, 부정적인 boolean보다는 긍정적인 boolean으로 명명하라

- 부정적이라면 두 번생각하게 되어 가독성을 해친다. 그래서 긍정적으로 바꾸었다.

pkeugine commented on src/main/java/model/Lotto.java:
```diff
@@ -36,18 +36,20 @@ private void validateSize(final List<LottoNumber> winningNumbers) {
     private void validateDuplicate(final List<LottoNumber> lottoNumbers) {
-        final Set<LottoNumber> set = new HashSet<>(lottoNumbers);
-        if (set.size() != lottoNumbers.size()) {
+        final Set<LottoNumber> removedDuplicate = new HashSet<>(lottoNumbers);
```
변

pkeugine commented on src/main/java/model/Lotto.java:
```diff
@@ -36,18 +36,20 @@ private void validateSize(final List<LottoNumber> winningNumbers) {
-        final Set<LottoNumber> set = new HashSet<>(lottoNumbers);
-        if (set.size() != lottoNumbers.size()) {
+        final Set<LottoNumber> removedDuplicate = new HashSet<>(lottoNumbers);
+        if (removedDuplicate.size() != lottoNumbers.size()) {
```
변수명을 removedDuplicate라고 지어주셨는데, “중복이 제거된 로또 번호” 라는 의미를 명확하게 나타내는 명사를 사용하면 어떨까요? 😊

pkeugine commented on src/main/java/model/Lotto.java:
```diff
@@ -36,18 +36,20 @@ private void validateSize(final List<LottoNumber> winningNumbers) {
     public int calculateWinningNumbersMatchCount(final WinningNumbers winningNumbers) {
-        return winningNumbers.calculateLottoMatchCount(getLottoNumbers());
+        return (int) lottoNumbers.stream()
+                .filter(winningNumbers::containsLottoNumber)
```
여기처럼 람다 표현식을 자주 활용하시는 것 같은데, 이 파일의 from 메서드 내 map 함수에서는 람다를 사용하지 않으셨네요! 🤔
물론 반드시 람다를 써야 하는 건 아니지만, 프로젝트 내에서 일관성을 유지하는 것이 코드 가독성과 유지보수성 측면에서 더 좋을 것 같아요.

changuii replied:
생성자라, 놓친 것 같습니다. 더욱 주의깊게 살펴보겠습니다.
반영했습니다. :)

pkeugine commented on src/test/java/model/LottoTest.java:
```diff
@@ -5,26 +5,31 @@
 
+    private List<LottoNumber> lottoNumbers;
+
+    @BeforeEach
```
`@BeforeEach` 활용 좋습니다!!👍👍

pkeugine commented on src/test/java/model/WinningNumbersTest.java:
```diff
@@ -6,70 +6,69 @@
+            WinningNumbers winningNumbers = new WinningNumbers(lottoNumbers, bonusBall);
 
             // then
 
```
여기 불필요한 줄바꿈이 보이네요! 👀

pkeugine commented on src/test/java/model/PurchaseAmountTest.java:
```diff
@@ -26,7 +26,7 @@ void createPurchaseAmount() {
 
             // then
-            assertThat(actual.getMoney()).isEqualTo(money);
+            assertThat(actual).extracting("money").isEqualTo(money);
```
이 부분에서는 생성자의 파라미터가 필드에 잘 들어갔는지를 `extracting()`으로 테스트하고 있는데, 꼭 생성자 테스트가 필요한지 한 번 고민해봐도 좋을 것 같아요! (아래 `createWinningNumbers` 테스트도 마찬가지고요.)

물론, 생성자 내부에 validate 로직이 있지만, 해당 부분은 아래 예외 케이스 테스트에서 충분히 검증하고 있어서 굳이 따로 테스트할 필요가 있을까 싶네요! 😃

changuii replied:
validate를 검증하는 로직이 있지만, 혹시 모를 **필드 초기화**를 검증한다고 생각했습니다.
저를 믿지 못해서 촘촘하게 했는데 현재는 불필요해보일지 몰라도 미래에는 필요할 수 있지 않을까요?

필드 초기화 검증에 대한 피케이의 생각이 궁금해요 🧐

pkeugine replied:
촘촘한 테스트 작성, 너무 좋습니다! 👍
현재는 불필요해 보일 수 있지만 나중에는 필요할 수도 있다는 점에서 충분히 고려할 만한 부분이에요.
저는 정적 팩토리 메서드를 통한 생성이라면 필드 초기화 검증이 의미가 있을 수 있다고 생각하지만,
생성자에서 변경 없이 바로 필드를 채우는 경우라면 생략해도 괜찮을 것 같아요.

changuii replied:
로직이 없다면 생략해도 된다는 말씀이시군요
감사합니다 :)

pkeugine commented on src/main/java/model/Lotto.java:
```diff
@@ -36,18 +36,20 @@ private void validateSize(final List<LottoNumber> winningNumbers) {
 
-    public boolean isContainsBonusNumber(final WinningNumbers winningNumbers) {
-        return winningNumbers.matchBonusNumber(getLottoNumbers());
+    public boolean isContainsNumber(final LottoNumber number) {
```
946e05940fb4ee396e8f6b712a0c65fd9f1c48b9 와 369f9ff7bae204cb13d5b6608f96641c957ce7ed 커밋의 변경내용 너무 좋습니다! 각 클래스의 역할을 고민해보신 것 같네요! 👍

pkeugine commented on src/main/java/factory/LottoFactory.java:
```diff
@@ -24,7 +24,10 @@ public static OutputView outputView() {
     }
 
-    public static NumberGenerator numberGenerator() {
+    public static RandomNumberGenerator numberGenerator() {
```
> `NumberGenerator` 라는 클래스 안에 `generateRandomNumber` 라는 메서드가 있는게 조금 어색해보여요!

저번 PR에서 해당 피드백을 드렸었죠! 이번에 클래스명에 `Random`을 추가해주셨는데, 그 과정에서 고민해볼 만한 부분이 생긴 것 같아요. `RandomNumberGenerator`와 `RandomUniqueNumberGenerator`의 차이가 단순히 `Unique` 여부뿐이라면, 이렇게 분리하는 게 의미가 있을까요? 🤔

테스트를 위해 추가된 `MockNumberGenerator`가 `RandomNumberGenerator`를 상속하면서도 랜덤하지 않은 숫자를 생성하는 역할을 하고 있는데, 조금 더 자연스러운 구조가 있을지 같이 고민해보면 좋을 것 같아요! 😊

changuii replied:
이에 대해 고민해보았고, 다음의 결론을 내렸습니다.
NumberGenerator라면, 굳이 generateRandomNumber라는 메서드가 상위 클래스에 존재하는 것이 부자연스럽다고 느꼈고, 그럼 이를 RandomUniqueNumberGenerator에서 구현해야 하는데, 그렇게된다면 추상 클래스의 의미가 희미해진다고 판단했습니다.
그래서 NumberGenerator interface를 만들어서, 추상화하여 사용하도록 반영하였습니다. :)

앞으로 좀 더 깊게 고민해보겠습니다.

pkeugine commented on src/main/java/constans/ErrorType.java:
```diff
@@ -2,21 +2,34 @@
     WINNING_RESULT_POSITIVE("당첨 결과는 음수일 수 없습니다."),
-    // input view
+
+    // ConsoleInputView
```
에러가 발생하는 위치가 분산되어 있는 것 같아요.

지금 보면 `WinningNumbers` 관련 에러 메시지 중에서 `WINNING_NUMBERS_INVALID_INPUT`은 `ConsoleInputView`에서 발생하는 반면, 나머지는 `WinningNumbers` 내부에서 발생하고 있네요. 🤔

실제로 실행해보면 "당첨 번호의 입력값이 올바르지 않습니다."는 당첨 번호를 입력한 직후에 발생하는데,
"당첨 번호는 6개여야 합니다."는 보너스볼을 입력한 후에 발생해서 흐름이 어색해지는 것 같아요.

이 부분을 조금 더 자연스럽게 조정할 방법이 있을까요? 고민해보면 좋을 것 같습니다! 😊

changuii replied:
저도 빠른 피드백? 동일한 시점에 예외가 발생하지 않는 것이 부자연스럽다고 느꼈지만 당첨 번호와 밀접한 연관이 있어서 그렇지 않을까BonusBall과 떨어지면 이것도 부자연스럽지 않을까? 고민해보았고 이렇게 작성했습니다.

일단 리뷰를 반영해서 BonusBall을 외부로 분리하여 작성했는데, 코멘트를 작성하면서 생각난 방법은 
WinningNumbers와 BonusBall 도메인을 합친, WinningLotto 클래스로 두 도메인을 감싸면 어떨까 라는 생각이 들었습니다. 

하지만, 연관되었다고해서 이걸 굳이 묶어야 할까라는 고민이 또 생기는데
피케이의 생각이 궁금합니다! 🧐

pkeugine replied:
Baseball 클래스를 분리해서 해결된 부분 확인했습니다.
`WinningNumbers와 BonusBall 도메인을 합친, WinningLotto 클래스로 두 도메인을 감싸는 것`도 해보세요! 
코드를 이리저리 바꿔보고, 각각의 장단점을 직접 경험하는 과정 자체가 의미 있는 작업이라고 생각합니다 😀

pkeugine commented on src/main/java/constans/ErrorType.java:
```diff
@@ -2,21 +2,34 @@
 
```
package 이름에 오타가 있네요!

pkeugine reviewed (changes requested):
안녕하세요 부기,
이번 단계도 잘 진행해주셨네요.
여유롭게 진행해주셔서 이번 스텝에서는 리뷰 반영 한 번 해보죠! :)
코멘트 확인해주시고, 이번엔 반영 또는 댓글을 모두 달면 다시 리뷰 요청을 남겨주세요.
남은 주말 잘 보내시고 파이팅입니다 🔥 🔥 🔥

changuii committed:
fix : constants 패키지 명의 오타를 수정하라

changuii committed:
refactor : WinningNumbersTest의 불필요한 줄바꿈을 제거하라

changuii committed:
refactor : 코드의 일관성을 위해 람다식으로 통일하라

changuii committed:
refactor : 변수명을 더욱 명확하게 작성하라

- removedDuplicate보다 명확하게 deduplicatedLottoNumber로 변경하여 가독성을 높였다.

changuii committed:
refactor : NumberGenerator의 역할을 더 자연스럽게 인터페이스로 추상화하여라

- 기존의 구조인, 추상 클래스가 random number 메서드를 가진것이 자연스럽지 않았다.
- 만약 상속받은 메서드가 없다면, 굳이 추상 클래스로 만들 필요가 없다고 생각해 인터페이스로 추상화하였다.
- MockNumberGenerator, RandomUniqueNumberGenerator로 더 자연스럽게 분리되었다.
- LottoMachine도 어떻게 번호가 추출되는지 알 필요 없이, 주입된 NumberGenerator로 번호만 뽑아 생성하면 되도록 하여 좀 더 자연스러워졌다.

changuii committed:
refactor : BonusBall 도메인을 만들어, WinningNumbers에서 보너스볼을 분리하라

- WinningNumbers의 예외 발생 시점을 Input 예외와 동일하게 하여 사용자에게 빠른 피드백을 제공하게 하였다.

changuii committed:
test : BonusBall에 대한 단위 테스트를 작성하라

changuii committed:
refactor : DisplayName 애노테이션의 위치를 Test 바로 하단에 위치하도록 하라

- 규칙을 일정하게 유지하여 가독성을 높이도록 했다.

changuii committed:
refactor : BonusBall의 TestFixture를 각 테스트마다 초기화되도록 하여라

- 각각의 테스트에 변경된 TestFixture가 영향을 끼치지 않도록 초기화되도록 했다.

changuii reviewed (commented):
정말 꼼꼼히 봐주셔서 감사합니다.!!

pkeugine commented on src/main/java/view/output/ConsoleOutputView.java:
```diff
@@ -20,28 +19,34 @@ public void printLottoNumbers(final List<List<Integer>> lottoNumbers) {
+        System.out.printf("총 수익률은 %.2f입니다.(기준이 1이기 때문에 결과적으로 %s라는 의미임)\n", revenueRate, revenueDescription);
+    }
+
+    private String selectByFlag(final boolean selectLeft, final String left, final String right) {
```
이 부분은 요구사항에 포함되지 않아서 따로 말씀드리진 않았는데요, `left` 와 `right` 라는 표현이 들어가면서 코드 읽기를 더 어렵게 만드는 요소가 추가되는 것 같아요. 🤔

특히 이익과 손해에 대한 기준이 잘못 적용된 것처럼 보입니다. `isRevenue = true`일 때 "손해"를 리턴하는 것이 맞는지 한 번 더 확인해보세요! 🤨

요구사항에 없는 기능을 추가할 때는 Readme 에 자세한 내용을 추가하는 것이 좋습니다!

changuii replied:
메서드의 변수명을 고민하다, left right가 직관적일 것이라 판단해서 작성했습니다. 하지만 어색하게 되는 것 같네요 ㅎㅎ
isDamage()라는 메서드가 부정적인 boolean을 반환하고 있어 한 번더 생각하게 되는 것을 방지하고자 isRevenue로 변경하면서 빠드린 것 같습니다. 
Readme에는 작성해두었습니다. 
두 부분 모두 반영해두었습니다. 꼼꼼히 봐주셔서 감사합니다 !! 😃

pkeugine reviewed (changes requested):
안녕하세요 부기!
단계를 거치며 많은 고민을 하고 계신 점 보기 좋습니다.
아직 수정하고자 하는 포인트가 몇 가지 있을 것 같으니
코멘트 한 번 확인해주세요!

changuii committed:
refactor : WinningNumbers와 BonusBall을 가진 WinningLotto를 만들어라

changuii committed:
test : 의미없는 this 키워드를 제거하라

changuii committed:
test : WinningLotto 도메인에 대한 단위 테스트를 작성하라

changuii committed:
fix : isRevenue, 이익이 발생한 경우 손해를, 손해가 발생한 경우 이익을 출력하는 문제를 해결하라

changuii committed:
refactor : selectByFlag 메서드의 변수명을 명확하게 변경하라

- left, right보다는 더 명확하게 trueValue, falseValue로 명명하였다.

changuii commented:
궁금증이 생겨 코멘트 남깁니다!!
현재 구조에서는 서비스가 커지면 커질수록 아래의 `controller.run()` 메서드가 커지는 문제가 생깁니다.

```java
    public void run() {
        final PurchaseAmount purchaseAmount = executeWithRetry(this::inputPurchaseAmount, 0);
        outputView.printPurchaseQuantity(purchaseAmount.calculateLottoCount());

        final LottoMachine lottoMachine = new LottoMachine(numberGenerator);
        final List<Lotto> lottos = lottoMachine.issueLottos(purchaseAmount);
        outputView.printLottoNumbers(convertLottos(lottos));

        final WinningLotto winningLotto = inputWinningLotto();

        final WinningResult winningResult = WinningResult.of(lottos, winningLotto);
        outputView.printLottoStatistics(winningResult.calculateRateOfRevenue(), winningResult.getLottoRanks(),
                winningResult.isRevenue());
    }
```

각각의 기능들은 다음 기능에서 사용할 필요가 있기 때문에 메서드로 추출하여 가질 수도 없었습니다.
이런 경우에는 어떤 식으로 분리해야하나요? 피케이의 생각이 궁금해요! 🧐

pkeugine commented:
> 궁금증이 생겨 코멘트 남깁니다!! 현재 구조에서는 서비스가 커지면 커질수록 아래의 `controller.run()` 메서드가 커지는 문제가 생깁니다.
> 
> ```java
>     public void run() {
>         final PurchaseAmount purchaseAmount = executeWithRetry(this::inputPurchaseAmount, 0);
>         outputView.printPurchaseQuantity(purchaseAmount.calculateLottoCount());
> 
>         final LottoMachine lottoMachine = new LottoMachine(numberGenerator);
>         final List<Lotto> lottos = lottoMachine.issueLottos(purchaseAmount);
>         outputView.printLottoNumbers(convertLottos(lottos));
> 
>         final WinningLotto winningLotto = inputWinningLotto();
> 
>         final WinningResult winningResult = WinningResult.of(lottos, winningLotto);
>         outputView.printLottoStatistics(winningResult.calculateRateOfRevenue(), winningResult.getLottoRanks(),
>                 winningResult.isRevenue());
>     }
> ```
> 
> 각각의 기능들은 다음 기능에서 사용할 필요가 있기 때문에 메서드로 추출하여 가질 수도 없었습니다. 이런 경우에는 어떤 식으로 분리해야하나요? 피케이의 생각이 궁금해요! 🧐

컨트롤러의 메서드가 커지는 문제를 벌써 고민하시는군요👍👍 
컨트롤러의 run 내부에서 메서드를 추출하거나 컨트롤러 자체를 분리하는 방향도 고려해볼 수 있을 것 같아요!

‘각각의 기능들은 다음 기능에서 사용할 필요가 있기 때문에 메서드로 추출하여 가질 수도 없었다’에 대해 답을 드리자면
구매 금액을 입출력 후 purchaseAmount를 리턴하는 메서드나 컨트롤러는 추출이 가능할 거 같아요. 
그렇게 추출하면 purchaseAmount의 입력과 출력 모두를 담당하는 부분이 생기겠죠?

이런 방법으로 나머지 부분도 분리할 수 있는지 고민해보면 좋을 것 같습니다. 물론 벌써 고민할 정도로 run이 긴 것은 아닙니다😀

pkeugine reviewed (approved):
수고 하셨습니다 👏👏
수정하신 부분들 잘 봤습니다. 첫 미션인데도 이것저것 열심히 고민하신게 보이네요! 
마지막 질문에 대한 코멘트 남기고 머지합니다. 
다음 미션도 파이팅!! 💪💪
